src/main/java/org/truffleruby/cext/CExtNodes.java

```diff
@@ -933,6 +937,65 @@ public DynamicObject threadWaitFDNode(int fd) {
 
     }
 
+    @CoreMethod(names = "rb_wait_for_single_fd", onSingleton = true, required = 4, lowerFixnum = {1, 2, 3, 4})
+    public abstract static class WaitForSingleFDNode extends CoreMethodArrayArgumentsNode {
+
+        private static final int RB_WAITFD_IN = 1;
+        private static final int RB_WAITFD_PRI = 2;
+        private static final int RB_WAITFD_OUT = 4;
+
+        @Specialization
+        public int waitForSingleFDNode(int fd, int events, long tv_secs, long tv_usecs) {
+            final boolean isRead = (events & RB_WAITFD_IN) != 0;
+            final boolean isWrite = (events & RB_WAITFD_OUT) != 0;
+            final boolean isError = (events & RB_WAITFD_PRI) != 0;
+            final FDSet readFdsSet = getFdSet(fd);
+            final FDSet writeFdsSet = getFdSet(fd);
+            final FDSet errorFdsSet = getFdSet(fd);
+            final jnr.ffi.Pointer readFds = isRead ? readFdsSet.getPointer() : Pointer.JNR_NULL;
+            final jnr.ffi.Pointer writeFds = isWrite ? writeFdsSet.getPointer() : Pointer.JNR_NULL;
+            final jnr.ffi.Pointer errorFds = isError ? errorFdsSet.getPointer() : Pointer.JNR_NULL;
+            final Timeval timeoutObject;
+            if (tv_secs > 0) {
+                timeoutObject = new DefaultNativeTimeval(jnr.ffi.Runtime.getSystemRuntime());
+                timeoutObject.setTime(new long[]{tv_secs, tv_usecs});
+            } else {
+                timeoutObject = null;
+            }
+            return getContext().getThreadManager().runBlockingSystemCallUntilResult(this, () -> {
+                final int fdsResult = nativeSockets().select(fd + 1,
+                        readFds,
+                        writeFds,
+                        errorFds,
+                        timeoutObject);
+                if(fdsResult < 0){
+                    final int errno = posix().errno();
+                    throw new RaiseException(ExceptionOperations.createSystemCallError(coreLibrary().getErrnoClass(Errno.valueOf(errno)), nil(), null, errno));
+                }
```

Comment:
You should check the return value outside runBlockingSystemCallUntilResult() if possible.
For instance, EINTR is a valid return value for it which automatically retries the blocking operation.